proposals/accepted/047-standard-library-reform.rst

```diff
@@ -0,0 +1,559 @@
+
+   In the new multi-library world, different libraries will inherit these special features, and we cannot be sure what the ramification will be until we try.
+
+   It is best to "practice" this by splitting ``base`` as soon as possible.
+   That will reduce the risk of everything else by both exploring "known unknowns" and scouting ahead for "unknown unknowns".
+
+#. Ultimately, in the name of rationalizing dependencies and the library division of labor, ``base`` will never make sense in anything like its current form.
+   We should therefore demote it to being a mere reexporter of other libraries that do make sense.
+
+**Step 1B**: MVP Split ``base`` by making it all reexports
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The first steps of `GHC issue #20647`_ track what needs to be done here.
+The key first step is finishing `GHC MR !7898`_.
+This is crude: a ``ghc-base`` that ``base`` merely reexports in full is just as ugly as the original ``base``, but this is the quickest route to de-risking the entire project as described in item 2 of the previous section.
+
+.. _`GHC issue #20647`: https://gitlab.haskell.org/ghc/ghc/-/issues/20647
+.. _`GHC MR !7898`: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7898
+
+**Step 2A**: Rationalize dependencies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At this point we have the following:
+
+- ``ghc-base``
+- ``base`` which reexports ``ghc-base``
+- A number of new libraries which reexport parts of ``base`` and possibly other libraries like ``text``.
+
+The goal is to shuffle code around so that we have something which makes more sense.
+That would look something like this:
+
+- 1 or more libraries in the GHC repo that are deeply tied to GHC's implementation details.
+  These libraries might depend on libraries in the next group.
+- 1 or more libraries outside the GHC that are repo agnostic to GHC's implementation details.
+  These libraries might depend on libraries in the previous group.
+- ``base``, lives in the GHC repo, and merely reexports functionality from the first two groups.
+- ``text``, lives outside the GHC repo, and should *not* depend on ``base``, but instead libraries from the first two groups.
+- The new standard libraries, living outside the GHC repo, merely reexporting functionality from the first two groups and possibly ``text``.
+
+It will take a while to untangle everything to get to this new maintainable end state.
+The good news is that we can get there very incrementally.
+The initial crude split will validate that shuffling definitions between libraries and modules works at all.
+After that, continuing to shuffle items reduces risk.
+
+The `GHC Wiki page on "Split Base" <https://gitlab.haskell.org/ghc/ghc/-/wikis/split-base>`_, especially Joachim Breitner's `prior attempt <https://github.com/nomeata/packages-base/blob/base-split/README.md>`_ offers good ideas backed by experience on where the natural cleavage points within ``base`` lie.
```

Comment:
This should have come earlier in the document.

Reply:
Good point, done in 491905de3ffb1060405a93091ac526271b363dd6